File: packages/grin/R/io.R
# io.R: the input contract for GRIN.
#
# Port of grintools' grin_io.py (Python). Same job, same principle: liberal about
# shape, strict about meaning. Any container is coerced, but two things are never
# guessed because guessing returns a confident wrong answer: (1) stimulus/response
# ORDER for a bare unlabelled matrix, and (2) counts vs proportions (the network
# reads trial totals as a second input, so proportions silently wreck the
# posterior's width). Both are refused unless the caller resolves them explicitly.
#
# Canonical layout: rows = stimuli, cols = responses, both in order A1B1, A1B2,
# A2B1, A2B2 (dimension A varies slowest, B fastest). A bare length-16 vector is
# read row-major (stimulus-major): position (0-based) k = stim_index*4 + resp_index.

#' Parse one cell label into 0-based (a_level, b_level)
#' @keywords internal
.parse_cell_label <- function(label, factor_a, factor_b, sep) {
  if (length(label) == 2) {
    a_name <- label[[1]]; b_name <- label[[2]]
  } else if (length(label) == 1) {
    parts <- strsplit(as.character(label), sep, fixed = TRUE)[[1]]
    if (length(parts) != 2) {
      stop(sprintf(
        "cannot split label '%s' on sep '%s' into two factor levels; pass a length-2 ",
        "c(a_name, b_name) or use a matching sep", label, sep), call. = FALSE)
    }
    a_name <- trimws(parts[1]); b_name <- trimws(parts[2])
  } else {
    stop(sprintf("label of length %d is not a 2-element pair or a single '%s'-separated string",
                 length(label), sep), call. = FALSE)
  }
  if (!(a_name %in% factor_a)) {
    stop(sprintf("'%s' is not a level of factor_a = c(%s)", a_name,
                 paste(factor_a, collapse = ", ")), call. = FALSE)
  }
  if (!(b_name %in% factor_b)) {
    stop(sprintf("'%s' is not a level of factor_b = c(%s)", b_name,
                 paste(factor_b, collapse = ", ")), call. = FALSE)
  }
  c(a = match(a_name, factor_a) - 1L, b = match(b_name, factor_b) - 1L)
}

#' Map 4 labels onto canonical order, or fail loudly
#'
#' Returns list(perm, placement). `perm` is a length-4 index vector into `labels`
#' such that `labels[perm[c]]` belongs at canonical position c. Raises unless the 4
#' labels form a complete 2x2 factorial (or are already the canonical tokens).
#' @keywords internal
.permutation_to_canonical <- function(labels, factor_a, factor_b, sep) {
  tok <- function(x) tolower(paste(x, collapse = sep))
  low <- vapply(labels, tok, character(1))
  if (identical(low, tolower(CANON_STIM)) || identical(low, tolower(CANON_RESP))) {
    placement <- setNames(labels, CANON_STIM)
    return(list(perm = 1:4, placement = placement))
  }
  if (is.null(factor_a) || is.null(factor_b)) {
    stop(paste("labels are not canonical A1B1.. tokens, so I need factor_a=c(A1,A2) and",
               "factor_b=c(B1,B2) to place them. Refusing to assume an order."), call. = FALSE)
  }
  factor_a <- as.character(factor_a); factor_b <- as.character(factor_b)
  source_of_canon <- rep(NA_integer_, 4)
  seen <- logical(4)
  placement <- vector("list", 4)
  for (pos in seq_along(labels)) {
    ab <- .parse_cell_label(labels[[pos]], factor_a, factor_b, sep)
    c_idx <- 2L * ab[["a"]] + ab[["b"]] + 1L
    if (seen[c_idx]) {
      stop(sprintf("label duplicates canonical cell %s; the 4 labels must be a complete 2x2 factorial",
                   CANON_STIM[c_idx]), call. = FALSE)
    }
    seen[c_idx] <- TRUE
    source_of_canon[c_idx] <- pos
    placement[[c_idx]] <- labels[[pos]]
  }
  if (!all(seen)) {
    stop(sprintf("labels do not cover the full factorial; missing: %s",
                 paste(CANON_STIM[!seen], collapse = ", ")), call. = FALSE)
  }
  names(placement) <- CANON_STIM
  list(perm = source_of_canon, placement = placement)
}

#' @keywords internal
.counts_guard <- function(counts, trials, warn) {
  counts <- matrix(as.numeric(counts), nrow = 4, ncol = 4)
  row_sums <- rowSums(counts)
  looks_like_props <- all(abs(row_sums - 1.0) < 1e-3) && all(counts <= 1.0 + .PROP_TOL)
  non_integer <- any(abs(counts - round(counts)) > .INT_TOL)
  if (looks_like_props || non_integer) {
    if (is.null(trials)) {
      stop(paste("input looks like PROPORTIONS, not counts (rows sum to ~1 or contain",
                 "non-integers). The network reads per-stimulus trial totals as a",
                 "separate input, so proportions would silently wreck the posterior",
                 "uncertainty. Pass counts, or pass trials=[...] to rescale."), call. = FALSE)
    }
    trials_vec <- as.numeric(trials)[1:4]
    counts <- round(sweep(counts, 1, trials_vec, `*`))   # row i scaled by trials[i]
    .warn(warn, paste("input treated as proportions and rescaled to counts using the",
                      "supplied trials; verify this is what you intended"))
  }
  matrix(as.integer(round(counts)), nrow = 4, ncol = 4)
}

#' @keywords internal
.row_warnings <- function(counts, trials, warn) {
  for (i in 1:4) {
    if (trials[i] < .SPARSE_TRIALS) {
      .warn(warn, sprintf("stimulus %s has only %d trials (< %d); posterior will be wide here",
                          CANON_STIM[i], as.integer(trials[i]), .SPARSE_TRIALS))
    }
  }
  if (any(counts == 0)) {
    zeros <- which(counts == 0, arr.ind = TRUE)
    zeros <- zeros[order(zeros[, "row"], zeros[, "col"]), , drop = FALSE]
    z_desc <- apply(zeros, 1, function(rc) sprintf("(%s, %s)", CANON_STIM[rc["row"]], CANON_RESP[rc["col"]]))
    .warn(warn, sprintf("%d empty cell(s) [%s]; fine for GRIN, but this is the cell-separation regime where MLE baselines diverge",
                        length(z_desc), paste(z_desc, collapse = ", ")))
  }
}

#' @keywords internal
.from_long <- function(data, factor_a, factor_b, sep, warn) {
  if (is.data.frame(data)) {
    cols <- tolower(names(data))
    if (!("stimulus" %in% cols) || !("response" %in% cols)) {
      stop("long data.frame needs 'stimulus' and 'response' columns (a 'count' column is optional)",
           call. = FALSE)
    }
    stim <- as.character(data[[which(cols == "stimulus")[1]]])
    resp <- as.character(data[[which(cols == "response")[1]]])
    cnt <- if ("count" %in% cols) as.numeric(data[[which(cols == "count")[1]]]) else rep(1, nrow(data))
  } else {
    stim <- vapply(data, function(r) as.character(r[[1]]), character(1))
    resp <- vapply(data, function(r) as.character(r[[2]]), character(1))
    cnt <- vapply(data, function(r) if (length(r) >= 3) as.numeric(r[[3]]) else 1, numeric(1))
  }
  stim_labels <- unique(stim); resp_labels <- unique(resp)
  if (length(stim_labels) != 4 || length(resp_labels) != 4) {
    stop(sprintf("expected 4 distinct stimuli and 4 responses; got %d stimuli, %d responses",
                 length(stim_labels), length(resp_labels)), call. = FALSE)
  }
  sp <- .permutation_to_canonical(as.list(stim_labels), factor_a, factor_b, sep)
  rp <- .permutation_to_canonical(as.list(resp_labels), factor_a, factor_b, sep)
  s_to_c <- setNames(integer(4), stim_labels[sp$perm]); s_to_c[stim_labels[sp$perm]] <- 1:4
  r_to_c <- setNames(integer(4), resp_labels[rp$perm]); r_to_c[resp_labels[rp$perm]] <- 1:4
  counts <- matrix(0, nrow = 4, ncol = 4)
  for (k in seq_along(stim)) {
    si <- s_to_c[[stim[k]]]; ri <- r_to_c[[resp[k]]]
    counts[si, ri] <- counts[si, ri] + cnt[k]
  }
  list(counts = counts, placement = sp$placement)
}

#' Normalise any supported input into a canonical-order confusion matrix
#'
#' Coerces a confusion matrix in (almost) any shape and label scheme into the
#' canonical-order (A1B1, A1B2, A2B1, A2B2 x A1B1, A1B2, A2B1, A2B2) form GRIN's
#' models expect, and never guesses the two things a wrong guess would silently
#' corrupt: stimulus/response order, and counts-vs-proportions.
#'
#' @param data A 4x4 matrix/data.frame, a length-16 vector (read row-major: stimulus
#'   varies slower than response), or (with `long = TRUE`) a long-format data.frame
#'   with `stimulus` and `response` columns (plus an optional `count` column).
#' @param stim_labels,resp_labels Character vectors of the 4 row/column labels, in
#'   the order `data` actually has them. If labels are canonical A1B1-style tokens
#'   they are matched directly; otherwise supply `factor_a`/`factor_b` too.
#' @param factor_a,factor_b Length-2 character vectors giving the two levels of each
#'   dimension, e.g. `c("Old","Young")`, `c("Neg","Pos")`, needed to place non-canonical
#'   labels like `"Old/Neg"`.
#' @param order Pass `"canonical"` to assert `data` is already in canonical order
#'   (skips label resolution entirely). A bare, unlabelled 4x4 with neither `order`
#'   nor labels is refused rather than guessed.
#' @param trials Optional per-stimulus trial totals (length 4); required if `data`
#'   looks like proportions rather than counts. Defaults to row sums of the counts.
#' @param sep Separator used to split single-string labels like `"Old/Neg"`.
#' @param long If `TRUE` (or `data` is a data.frame with a `stimulus` column),
#'   treat `data` as long-format trial-level or aggregated data.
#' @return A `grin_confusion_input` object: `$counts` (4x4 integer matrix), `$trials`
#'   (length-4 integer), `$placement`, `$warnings`, `$asserted_order`.
#' @export
grin_to_confusion <- function(data, stim_labels = NULL, resp_labels = NULL,
                              factor_a = NULL, factor_b = NULL, order = NULL,
                              trials = NULL, sep = "/", long = FALSE) {
  warn <- .new_warn_collector()
  asserted <- FALSE

  is_long <- isTRUE(long) || (is.data.frame(data) && "stimulus" %in% tolower(names(data)))
  if (is_long) {
    parsed <- .from_long(data, factor_a, factor_b, sep, warn)
    counts <- parsed$counts; placement <- parsed$placement
  } else {
    if (is.data.frame(data)) {
      if (is.null(resp_labels)) resp_labels <- names(data)
      idx <- rownames(data)
      if (is.null(stim_labels) && !all(grepl("^[0-9]+$", idx))) stim_labels <- idx
      mat <- matrix(as.numeric(as.matrix(data)), nrow = 4, ncol = 4)
    } else if (!is.null(dim(data)) && length(dim(data)) == 2) {
      mat <- matrix(as.numeric(data), nrow = 4, ncol = 4)
    } else {
      v <- as.numeric(data)
      if (length(v) != 16) {
        stop(sprintf(paste("expected a 4x4 matrix/data.frame or a length-16 vector",
                           "(read row-major: stimulus-major); got length %d"), length(v)),
             call. = FALSE)
      }
      mat <- matrix(v, nrow = 4, ncol = 4, byrow = TRUE)
    }

    if (identical(order, "canonical")) {
      counts <- mat
      placement <- setNames(as.list(CANON_STIM), CANON_STIM)
      asserted <- TRUE
    } else if (!is.null(stim_labels) || !is.null(resp_labels)) {
      if (!is.null(resp_labels) && is.null(stim_labels)) {
        stim_labels <- resp_labels
        .warn(warn, "rows were unlabelled; assumed to follow the same category order as the column labels")
      }
      if (is.null(stim_labels) || is.null(resp_labels)) {
        stop("need both stim_labels and resp_labels (or order='canonical') to place a 4x4 matrix",
             call. = FALSE)
      }
      rp <- .permutation_to_canonical(as.list(stim_labels), factor_a, factor_b, sep)
      cp <- .permutation_to_canonical(as.list(resp_labels), factor_a, factor_b, sep)
      counts <- mat[rp$perm, cp$perm, drop = FALSE]
      placement <- rp$placement
    } else {
      stop(paste("a bare 4x4 with no labels and no order assertion is refused: I will",
                 "not guess the stimulus/response order and hand back a confident wrong",
                 "posterior. Either pass order='canonical' to assert your matrix is",
                 "already A1B1,A1B2,A2B1,A2B2, or pass stim_labels/resp_labels with",
                 "factor_a and factor_b."), call. = FALSE)
    }
  }

  counts <- .counts_guard(counts, trials, warn)
  resolved_trials <- if (!is.null(trials)) {
    as.integer(round(as.numeric(trials)))[1:4]
  } else {
    as.integer(rowSums(counts))
  }
  if (!is.null(trials) && !identical(resolved_trials, as.integer(rowSums(counts)))) {
    .warn(warn, "supplied trials disagree with row sums of the counts; using the supplied trials")
  }
  .row_warnings(counts, resolved_trials, warn)

  structure(list(counts = counts, trials = resolved_trials, placement = placement,
                warnings = warn$msgs, asserted_order = asserted),
           class = "grin_confusion_input")
}

#' @export
print.grin_confusion_input <- function(x, ...) {
  cat(sprintf("<grin_confusion_input> trials=[%s] warnings=%d\n",
             paste(x$trials, collapse = ", "), length(x$warnings)))
  invisible(x)
}

#' Describe how input would be parsed, without running the network
#'
#' Setup guide / dev mode / exception surface in one. Never raises: parse errors are
#' captured and printed, not thrown, so this is safe to run while wiring up an
#' experiment. Pass the same arguments you would pass to [grin_to_confusion()].
#'
#' @inheritParams grin_to_confusion
#' @param printout Print a human-readable report (default `TRUE`).
#' @param ... Passed on to [grin_to_confusion()] (`stim_labels`, `factor_a`, etc.).
#' @return Invisibly, a list report (`$ready`, `$errors`, `$warnings`, `$counts`,
#'   `$trials`, `$placement`, `$asserted_order`).
#' @export
grin_describe <- function(data, printout = TRUE, ...) {
  report <- list(ready = FALSE, errors = character(0), warnings = character(0),
                 counts = NULL, trials = NULL, placement = NULL, asserted_order = NULL)
  ci <- tryCatch(grin_to_confusion(data, ...), error = function(e) e)
  if (inherits(ci, "grin_confusion_input")) {
    report$ready <- TRUE
    report$counts <- ci$counts; report$trials <- ci$trials
    report$placement <- ci$placement; report$warnings <- ci$warnings
    report$asserted_order <- ci$asserted_order
  } else {
    report$errors <- c(report$errors, sprintf("%s: %s", class(ci)[1], conditionMessage(ci)))
  }
  if (isTRUE(printout)) {
    lines <- c("GRIN input check", strrep("-", 52))
    if (report$ready) {
      lines <- c(lines, paste0("parsed OK, ready for inference",
                if (isTRUE(report$asserted_order)) "  (order asserted by caller)"
                else "  (order resolved from labels)"))
      lines <- c(lines, "canonical placement (canonical cell <- your label):")
      for (cnm in CANON_STIM) {
        lines <- c(lines, sprintf("    %s <- %s", cnm, paste(report$placement[[cnm]], collapse = "/")))
      }
      lines <- c(lines, paste("trials per stimulus:", paste(report$trials, collapse = ", ")))
      lines <- c(lines, "counts (canonical order):")
      for (i in 1:4) {
        lines <- c(lines, sprintf("    %-5s | %s", CANON_STIM[i],
                                  paste(sprintf("%4d", report$counts[i, ]), collapse = " ")))
      }
    } else {
      lines <- c(lines, "NOT ready, could not parse:")
    }
    for (w in report$warnings) lines <- c(lines, sprintf("  warning: %s", w))
    for (e in report$errors) lines <- c(lines, sprintf("  ERROR:   %s", e))
    cat(paste(lines, collapse = "\n"), "\n")
  }
  invisible(report)
}

File: packages/grintools/grintools/io.py
"""
grin_io.py: the input contract and stopping decisions for GRIN.

This is the thin layer between an experiment and the trained network. It does
three jobs and nothing else:

    NORMALISE   whatever the experiment has (labelled 4x4, bare 4x4, long trial
                log, aggregated long) -> the canonical-order 4x4 count matrix plus
                per-stimulus trial totals that api.infer expects.
    DESCRIBE    echo back exactly what was parsed, WITHOUT running the network, so
                an experimenter can confirm their wiring before trusting a result.
    DECIDE      turn a posterior (and, optionally, the construct probabilities from
                model_posterior) into a stop / keep-collecting decision, using a
                criterion the EXPERIMENTER declares. GRIN does not choose the rule.

Principle: liberal about shape, strict about meaning. Any container/layout is
coerced, but two things are never guessed because guessing returns a confident
wrong answer: (1) stimulus/response ORDER for a bare unlabelled matrix, and (2)
counts vs proportions (the network reads trial totals as a second input, so
proportions silently wreck the posterior's width). Both are refused unless the
caller resolves them explicitly.

Canonical layout (from grt_model.py): rows = stimuli, cols = responses, both in
order A1B1, A1B2, A2B1, A2B2 (dimension A = x, B = y, B fastest). Canonical index
of a cell = 2*a_level + b_level.

Australian English. numpy required; pandas optional (duck-typed).
"""

from __future__ import annotations
import numpy as np

# --------------------------------------------------------------------------- #
# Canonical layout (mirror grt_model.py; fall back so this runs standalone).
# --------------------------------------------------------------------------- #
try:
    from grt_model import STIMULUS_ORDER as CANON_STIM, RESPONSE_ORDER as CANON_RESP
    from grt_model import PARAM_NAMES
except Exception:
    CANON_STIM = ["A1B1", "A1B2", "A2B1", "A2B2"]
    CANON_RESP = ["a1b1", "a1b2", "a2b1", "a2b2"]
    PARAM_NAMES = ([f"zx_{i}" for i in range(4)]
                   + [f"zy_{i}" for i in range(4)]
                   + [f"rho_{i}" for i in range(4)])

PARAM_GROUPS = {"zx": list(range(0, 4)), "zy": list(range(4, 8)), "rho": list(range(8, 12))}

_SPARSE_TRIALS = 20
_PROP_TOL = 1e-6
_INT_TOL = 1e-8


# =========================================================================== #
# Containers
# =========================================================================== #
class ConfusionInput:
    """A normalised, network-ready input plus a record of how it was parsed."""
    def __init__(self, counts, trials, placement, warnings, asserted_order):
        self.counts = counts                  # (4,4) int, canonical order
        self.trials = trials                  # (4,)  int, per-stimulus totals
        self.placement = placement            # canonical cell -> source label
        self.warnings = warnings              # list[str]
        self.asserted_order = asserted_order  # did the caller assert canonical order?

    @property
    def ready(self):
        return self.counts is not None

    def __repr__(self):
        return (f"ConfusionInput(trials={list(map(int, self.trials))}, "
                f"warnings={len(self.warnings)})")


# =========================================================================== #
# Label resolution (strict): map 4 labels onto canonical order, or fail loudly.
# =========================================================================== #
def _canonical_index(a_level, b_level):
    return 2 * int(a_level) + int(b_level)


def _parse_cell_label(label, factor_a, factor_b, sep):
    if isinstance(label, (tuple, list)):
        if len(label) != 2:
            raise ValueError(f"label {label!r} is not a 2-tuple (a_name, b_name)")
        a_name, b_name = label
    else:
        parts = str(label).split(sep)
        if len(parts) != 2:
            raise ValueError(
                f"cannot split label {label!r} on sep {sep!r} into two factor "
                f"levels; pass a (a_name, b_name) tuple or use a matching sep")
        a_name, b_name = parts[0].strip(), parts[1].strip()
    if a_name not in factor_a:
        raise ValueError(f"'{a_name}' from {label!r} is not a level of factor_a={factor_a}")
    if b_name not in factor_b:
        raise ValueError(f"'{b_name}' from {label!r} is not a level of factor_b={factor_b}")
    return factor_a.index(a_name), factor_b.index(b_name)


def _permutation_to_canonical(labels, factor_a, factor_b, sep):
    """Return (perm, placement) so axis[perm[c]] is canonical cell c. Raises unless
    the 4 labels form a complete 2x2 factorial."""
    canon_tokens = [t.lower() for t in CANON_STIM]
    low = [str(x).lower() for x in labels]
    if low == canon_tokens or low == [t.lower() for t in CANON_RESP]:
        return list(range(4)), {CANON_STIM[i]: labels[i] for i in range(4)}
    if factor_a is None or factor_b is None:
        raise ValueError(
            "labels are not canonical A1B1.. tokens, so I need factor_a=(A1,A2) and "
            "factor_b=(B1,B2) to place them. Refusing to assume an order.")
    factor_a, factor_b = tuple(factor_a), tuple(factor_b)
    source_of_canon, seen = {}, set()
    for pos, lab in enumerate(labels):
        a, b = _parse_cell_label(lab, factor_a, factor_b, sep)
        c = _canonical_index(a, b)
        if c in seen:
            raise ValueError(f"label {lab!r} duplicates canonical cell {CANON_STIM[c]}; "
                             f"the 4 labels must be a complete 2x2 factorial")
        seen.add(c); source_of_canon[c] = pos
    missing = [CANON_STIM[c] for c in range(4) if c not in seen]
    if missing:
        raise ValueError(f"labels do not cover the full factorial; missing: {missing}")
    perm = [source_of_canon[c] for c in range(4)]
    placement = {CANON_STIM[c]: labels[source_of_canon[c]] for c in range(4)}
    return perm, placement


# =========================================================================== #
# Normaliser
# =========================================================================== #
def _is_dataframe(x):
    return hasattr(x, "columns") and hasattr(x, "itertuples") and hasattr(x, "index")


def _counts_guard(counts, trials, warnings):
    counts = np.asarray(counts, dtype=float)
    row_sums = counts.sum(axis=1)
    looks_like_props = np.allclose(row_sums, 1.0, atol=1e-3) and np.all(counts <= 1.0 + _PROP_TOL)
    non_integer = np.any(np.abs(counts - np.round(counts)) > _INT_TOL)
    if looks_like_props or non_integer:
        if trials is None:
            raise ValueError(
                "input looks like PROPORTIONS, not counts (rows sum to ~1 or contain "
                "non-integers). The network reads per-stimulus trial totals as a "
                "separate input, so proportions would silently wreck the posterior "
                "uncertainty. Pass counts, or pass trials=[...] to rescale.")
        trials = np.asarray(trials, dtype=float).reshape(4)
        counts = np.round(counts * trials[:, None])
        warnings.append("input treated as proportions and rescaled to counts using the "
                        "supplied trials; verify this is what you intended")
    return np.round(counts).astype(int)


def _row_warnings(counts, trials, warnings):
    for i, n in enumerate(trials):
        if n < _SPARSE_TRIALS:
            warnings.append(f"stimulus {CANON_STIM[i]} has only {int(n)} trials "
                            f"(< {_SPARSE_TRIALS}); posterior will be wide here")
    if np.any(counts == 0):
        z = [(CANON_STIM[i], CANON_RESP[j]) for i in range(4) for j in range(4)
             if counts[i, j] == 0]
        warnings.append(f"{len(z)} empty cell(s) {z}; fine for GRIN, but this is the "
                        f"cell-separation regime where MLE baselines diverge")


def _from_long(data, factor_a, factor_b, sep, warnings):
    stim, resp, cnt = [], [], []
    if _is_dataframe(data):
        cols = [str(c).lower() for c in data.columns]
        def col(name): return data.columns[cols.index(name)]
        if "stimulus" not in cols or "response" not in cols:
            raise ValueError("long DataFrame needs 'stimulus' and 'response' columns "
                             "(a 'count' column is optional)")
        has_count = "count" in cols
        for row in data.itertuples(index=False):
            d = row._asdict()
            stim.append(d[col("stimulus")]); resp.append(d[col("response")])
            cnt.append(d[col("count")] if has_count else 1)
    else:
        for row in data:
            row = tuple(row)
            if len(row) == 2:
                stim.append(row[0]); resp.append(row[1]); cnt.append(1)
            elif len(row) == 3:
                stim.append(row[0]); resp.append(row[1]); cnt.append(row[2])
            else:
                raise ValueError(f"long row {row!r} must be (stimulus, response) or "
                                 f"(stimulus, response, count)")
    stim_labels = list(dict.fromkeys(stim))
    resp_labels = list(dict.fromkeys(resp))
    if len(stim_labels) != 4 or len(resp_labels) != 4:
        raise ValueError(f"expected 4 distinct stimuli and 4 responses; got "
                         f"{len(stim_labels)} stimuli, {len(resp_labels)} responses")
    s_perm, s_place = _permutation_to_canonical(stim_labels, factor_a, factor_b, sep)
    r_perm, _ = _permutation_to_canonical(resp_labels, factor_a, factor_b, sep)
    s_to_c = {stim_labels[s_perm[c]]: c for c in range(4)}
    r_to_c = {resp_labels[r_perm[c]]: c for c in range(4)}
    counts = np.zeros((4, 4), dtype=float)
    for s, r, c in zip(stim, resp, cnt):
        counts[s_to_c[s], r_to_c[r]] += c
    return counts, s_place


def to_confusion(data, *, stim_labels=None, resp_labels=None, factor_a=None,
                 factor_b=None, order=None, trials=None, sep="/", long=False):
    """Normalise any supported input into a canonical-order ConfusionInput.
    Raises ValueError on any ambiguity of meaning."""
    warnings = []
    factor_a = tuple(factor_a) if factor_a is not None else None
    factor_b = tuple(factor_b) if factor_b is not None else None

    is_long = long or (_is_dataframe(data)
                       and "stimulus" in [str(c).lower() for c in data.columns])
    if is_long:
        counts, placement = _from_long(data, factor_a, factor_b, sep, warnings)
        asserted = False
    else:
        if _is_dataframe(data):
            if resp_labels is None:
                resp_labels = list(data.columns)
            idx = list(data.index)
            if stim_labels is None and not all(isinstance(i, (int, np.integer)) for i in idx):
                stim_labels = idx
            mat = np.asarray(data.values, dtype=float)
        else:
            mat = np.asarray(data, dtype=float)
        if mat.shape != (4, 4):
            mat = mat.reshape(4, 4)

        if order == "canonical":
            counts = mat
            placement = {c: c for c in CANON_STIM}
            asserted = True
        elif stim_labels is not None or resp_labels is not None:
            if resp_labels is not None and stim_labels is None:
                stim_labels = list(resp_labels)
                warnings.append("rows were unlabelled; assumed to follow the same "
                                "category order as the column labels")
            if stim_labels is None or resp_labels is None:
                raise ValueError("need both stim_labels and resp_labels (or "
                                 "order='canonical') to place a 4x4 matrix")
            r_perm, placement = _permutation_to_canonical(stim_labels, factor_a, factor_b, sep)
            c_perm, _ = _permutation_to_canonical(resp_labels, factor_a, factor_b, sep)
            counts = mat[np.ix_(r_perm, c_perm)]
            asserted = False
        else:
            raise ValueError(
                "a bare 4x4 with no labels and no order assertion is refused: I will "
                "not guess the stimulus/response order and hand back a confident wrong "
                "posterior. Either pass order='canonical' to assert your matrix is "
                "already A1B1,A1B2,A2B1,A2B2, or pass stim_labels/resp_labels with "
                "factor_a and factor_b.")

    counts = _counts_guard(counts, trials, warnings)
    resolved_trials = (np.asarray(trials, dtype=float).round().astype(int).reshape(4)
                       if trials is not None else counts.sum(axis=1))
    if trials is not None and not np.array_equal(resolved_trials, counts.sum(axis=1)):
        warnings.append("supplied trials disagree with row sums of the counts; using "
                        "the supplied trials")
    _row_warnings(counts, resolved_trials, warnings)
    return ConfusionInput(counts, resolved_trials, placement, warnings, asserted)


# =========================================================================== #
# Describe: setup guide / dev mode / exception surface in one. Never raises.
# =========================================================================== #
def describe(data, *, printout=True, **kwargs):
    report = {"ready": False, "errors": [], "warnings": [], "counts": None,
              "trials": None, "placement": None, "asserted_order": None}
    try:
        ci = to_confusion(data, **kwargs)
        report.update(ready=True, counts=ci.counts, trials=ci.trials,
                      placement=ci.placement, warnings=ci.warnings,
                      asserted_order=ci.asserted_order)
    except Exception as e:
        report["errors"].append(f"{type(e).__name__}: {e}")
    if printout:
        lines = ["GRIN input check", "-" * 52]
        if report["ready"]:
            lines.append("parsed OK, ready for inference"
                         + ("  (order asserted by caller)" if report["asserted_order"]
                            else "  (order resolved from labels)"))
            lines.append("canonical placement (canonical cell <- your label):")
            for c in CANON_STIM:
                lines.append(f"    {c} <- {report['placement'][c]!r}")
            lines.append(f"trials per stimulus: {list(map(int, report['trials']))}")
            lines.append("counts (canonical order):")
            for i, rowc in enumerate(report["counts"]):
                lines.append(f"    {CANON_STIM[i]:5s} | " + " ".join(f"{v:4d}" for v in rowc))
        else:
            lines.append("NOT ready, could not parse:")
        for w in report["warnings"]:
            lines.append(f"  warning: {w}")
        for e in report["errors"]:
            lines.append(f"  ERROR:   {e}")
        print("\n".join(lines))
    return report


# =========================================================================== #
# Response bias: computed directly from a confusion matrix, no model fit
# required. GRT's separability/independence machinery lives entirely in
# infer()'s 12 identified parameters; the raw tendency to over- or
# under-report one level of a dimension, independent of how well the observer
# discriminates it, is a property of the data alone and worth a name.
# =========================================================================== #
def response_bias(counts, trials=None):
    """Response bias from a raw confusion matrix.

    The signed tendency to report level 2 of a dimension more or less often
    than level 1, averaged across the four stimuli: 0 is unbiased, positive
    means the observer favours the "2" response on that dimension more than a
    fair coin would, negative means they favour "1". This describes the data,
    independent of infer()'s model fit -- it needs no trained network and
    works even on a matrix GRIN can't otherwise fit.

    `counts`: a canonical-order 4x4 matrix, or a length-16 vector read
    row-major. `trials`: optional per-stimulus trial totals; defaults to row
    sums.

    Returns a dict: `x_bias`, `y_bias` (each in [-0.5, 0.5]), and `p_resp2`,
    a (4, 2) array giving P(respond level 2) on each dimension for each of
    the four stimuli (so a systematic bias can be told apart from one driven
    by a single stimulus).
    """
    cm = np.asarray(counts, dtype=float).reshape(4, 4)
    if trials is None:
        trials = cm.sum(axis=1)
    trials = np.asarray(trials, dtype=float).reshape(4)
    props = cm / trials[:, None]
    p_x2 = props[:, 2] + props[:, 3]   # respond "A2" (canonical cols a2b1, a2b2)
    p_y2 = props[:, 1] + props[:, 3]   # respond "B2" (canonical cols a1b2, a2b2)
    return {
        "x_bias": float(p_x2.mean() - 0.5),
        "y_bias": float(p_y2.mean() - 0.5),
        "p_resp2": np.stack([p_x2, p_y2], axis=1),
    }


# =========================================================================== #
# Convenience: normalise -> infer (-> constructs -> decide) in one call. Network
# imports are lazy so this module and its I/O are usable without torch present.
# =========================================================================== #
def _import_infer():
    try:
        from api import infer, load_model
    except Exception:
        from .api import infer, load_model
    return infer, load_model


def _import_model_posterior():
    try:
        from model_posterior import model_posterior
    except Exception:
        from .model_posterior import model_posterior
    return model_posterior


def fit(data, *, stim_labels=None, resp_labels=None, factor_a=None, factor_b=None,
        order=None, trials=None, sep="/", long=False, criterion=None,
        constructs=False, model=None, n_samples=1000):
    """Normalise `data`, run GRIN, optionally compute construct probabilities and a
    stopping decision. Returns the InferenceResult with `.input_` attached, plus
    `.constructs` and `.decision` when requested."""
    ci = to_confusion(data, stim_labels=stim_labels, resp_labels=resp_labels,
                      factor_a=factor_a, factor_b=factor_b, order=order,
                      trials=trials, sep=sep, long=long)
    infer, load_model = _import_infer()
    want_constructs = constructs or (criterion is not None and criterion.needs_constructs)
    if model is None and want_constructs:
        model = load_model()                      # load once, share with model_posterior
    result = infer(ci.counts, trials=ci.trials, model=model, n_samples=n_samples)
    result.input_ = ci
    if want_constructs:
        model_posterior = _import_model_posterior()
        result.constructs = model_posterior(
            model, ci.counts.reshape(1, 16).astype(float),
            ci.trials.reshape(1, 4).astype(float), n_samples=n_samples)[0]
    if criterion is not None:
        result.decision = criterion.evaluate(result, getattr(result, "constructs", None))
    return result
